### PR TITLE
Add support for wasm runtime metrics try #2

### PR DESCRIPTION
**Pull request by sandreim**
Fixes https://github.com/paritytech/polkadot/issues/4383

Companion for https://github.com/paritytech/substrate/pull/10440

skip check-dependent-cumulus

### CODE REVIEW

**Review by drahnr — COMMENTED** [review]
I like the approach with a second order function :+1:

File: node/metrics/src/runtime.rs [hunk 1]
```diff
@@ -108,7 +108,8 @@ impl RuntimeMetricsProvider {
 impl sc_tracing::TraceHandler for RuntimeMetricsProvider {
 	fn handle_span(&self, _span: &sc_tracing::SpanDatum) {}
 	fn handle_event(&self, event: &sc_tracing::TraceEvent) {
-		if event.target.ne("metrics") {
+		let target = event.values.string_values.get("target");
+		if target.is_none() || target.unwrap().ne("metrics") {
```
Comment (drahnr) [inline node/metrics/src/runtime.rs:112]: Use a `map_or` call instead

File: cli/src/cli.rs [hunk 2]
```diff
@@ -121,3 +121,11 @@ pub struct Cli {
...
 	pub run: RunCmd,
 }
+
+impl Cli {
+	/// Updates the prometheus metric prefix.
+	/// Must be called before `CliConfiguration::create_configuration()`.
+	pub fn update_prometheus_metric_prefix(&mut self, prefix: &'static str) {
+		self.run.base.prometheus_metric_prefix = prefix;
+	}
+}
```
Comment (drahnr) [inline cli/src/cli.rs:131]: Not really a fan of modifying the `Cli` struct.

Comment (sandreim) [reply cli/src/cli.rs:131]: This is not needed anymore as the metrics prefix has been removed in substrate.

Comment (sandreim) [reply cli/src/cli.rs]: removed

File: primitives/src/v0.rs [hunk 3]
```diff
@@ -923,6 +923,74 @@ pub mod fisherman {
 	}
 }
 
+// TODO: add feature gate.
+/// Runtime metric operations.
+#[derive(Debug, PartialEq, Eq, Clone, Encode, Decode, TypeInfo)]
```
Comment (sandreim) [inline primitives/src/v0.rs:928]: I don't think all of these are needed.

Comment (drahnr) [reply primitives/src/v0.rs:928]: `v0` is deprecated and will be removed soon™

Comment (sandreim) [reply primitives/src/v0.rs:928]: Thanks. Got it.

File: node/metrics/src/runtime.rs [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+	}
+
+	/// Register a counter vec metric.
+	pub fn register_countervec(
+		&self,
+		metric_name: &str,
+		description: &str,
+		label: RuntimeMetricLabel,
+	) -> Result<(), PrometheusError> {
+		if !self.1.counter_vecs.lock().expect("bad lock").contains_key(metric_name) {
```
Comment (drahnr) [inline node/metrics/src/runtime.rs:52]: Hold the lock for the entire scope

Comment (drahnr) [reply node/metrics/src/runtime.rs:52]: Also, runtime never panics, handle it and print a warning

File: node/metrics/src/runtime.rs [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+			.string_values
+			.get("target")
+			.unwrap_or(&String::default())
+			.ne("metrics")
+		{
+			return
+		}
+
+		if let Some(update_op_str) = event.values.string_values.get("params").cloned() {
+			// TODO: Fix ugly hack because the payload comes in as a formatted string.
```
Comment (drahnr) [inline node/metrics/src/runtime.rs:104]: As said, adding a base64 encoding would avoid this entirely.

File: runtime/parachains/Cargo.toml [hunk 6]
```diff
@@ -23,6 +23,7 @@ sp-session = { git = "https://github.com/paritytech/substrate", branch = "master
 sp-staking = { git = "https://github.com/paritytech/substrate", branch = "master", default-features = false }
 sp-core = { git = "https://github.com/paritytech/substrate", branch = "master", default-features = false }
 sp-keystore = { git = "https://github.com/paritytech/substrate", branch = "master", optional = true }
+sp-tracing = { version = "4.0.0-dev", branch = "master", git = "https://github.com/paritytech/substrate", default-features = false }
```
Comment (drahnr) [inline runtime/parachains/Cargo.toml:26]: These dependencies should be `optional = true`

File: runtime/parachains/Cargo.toml [hunk 7]
```diff
@@ -42,6 +43,7 @@ primitives = { package = "polkadot-primitives", path = "../../primitives", defau
 
 rand = { version = "0.8.3", default-features = false }
 rand_chacha = { version = "0.3.1", default-features = false }
+polkadot-runtime-metrics = { path = "../metrics", default-features = false}
```
Comment (drahnr) [inline runtime/parachains/Cargo.toml:46]: CC

File: runtime/parachains/src/paras_inherent/mod.rs [hunk 8]
```diff
@@ -374,6 +382,7 @@ impl<T: Config> Pallet<T> {
 			disputed_bitfield
 		};
 
+		bitfields_processed_metric.inc_by(signed_bitfields.len().try_into().unwrap_or_default());
```
Comment (drahnr) [inline runtime/parachains/src/paras_inherent/mod.rs:385]: I don't get the `try_into().unwrap_or_default()`. Shouldn't this be a `let _ = ` to silence the unused results?

File: runtime/metrics/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Runtime metric interface similar to native Prometheus metrics.
```
Comment (drahnr) [inline runtime/metrics/src/lib.rs:17]: Add a note that this is only meant for debugging and should never end up in production releases.

File: primitives/src/v0.rs [hunk 10]
```diff
@@ -923,6 +923,74 @@ pub mod fisherman {
...
+	/// The name of the metric.
+	pub metric_name: Vec<u8>,
+	/// The operation applied to the metric.
+	pub op: RuntimeMetricOp,
+}
+
+impl RuntimeMetricUpdate {
+	/// Returns the metric name.
+	pub fn metric_name(&self) -> &str {
+		unsafe { sp_std::str::from_utf8_unchecked(&self.metric_name) }
```
Comment (drahnr) [inline primitives/src/v0.rs:946]: I am not a fan of this. Can't we use `String::parse_from_utf8_lossy` in the std case?

**Review by drahnr — CHANGES_REQUESTED** [review]
General approach looks good :+1: , a few nits and some unfinished ends to tie up :)

File: node/metrics/src/runtime.rs (hunk 5 above)
Comment (sandreim) [reply node/metrics/src/runtime.rs:104]: Added bs58 encoding.

File: node/metrics/src/runtime.rs [hunk 11]
```diff
@@ -126,6 +128,23 @@ impl RuntimeMetricsProvider {
 			},
 		}
 	}
+
+	// Returns the `bs58` encoded metric update operation.
+	fn parse_event_params(event_params: &String) -> Vec<u8> {
+		println!("params: {}", event_params);
+
+		// Shave " }" suffix.
+		let new_len = event_params.len() - 2;
```
Comment (drahnr) [inline node/metrics/src/runtime.rs:137]: The range calculations are all a bit too optimistic for my taste, and should assume too short or changed prefixes and hence use saturating arthmetic

File: runtime/parachains/src/paras_inherent/mod.rs (hunk 8 above)
Comment (sandreim) [reply runtime/parachains/src/paras_inherent/mod.rs:385]: There is no unused result in this case, it's converting the value to u64.

File: node/metrics/Cargo.toml [hunk 12]
```diff
@@ -8,10 +8,18 @@ description = "Subsystem traits and message definitions"
...
 metered-channel = { path = "../metered-channel" }
 
 substrate-prometheus-endpoint = { git = "https://github.com/paritytech/substrate", branch = "master" }
+sc-tracing = { git = "https://github.com/paritytech/substrate", branch = "master" }
+sc-cli = { git = "https://github.com/paritytech/substrate", branch = "master" }
+sc-service = { git = "https://github.com/paritytech/substrate", branch = "master" }
+codec = { package = "parity-scale-codec", version = "2.2.0" }
+sp-tracing = { git = "https://github.com/paritytech/substrate", branch = "master" }
+primitives = { package = "polkadot-primitives", path = "../../primitives/" }
+bs58 = { version = "0.4.0", features = ["alloc"] }
```
Comment (drahnr) [inline node/metrics/Cargo.toml:22]: Do we really all of these deps, especially `sc-service` and `sc-cli` are particularly heavy.

File: node/metrics/Cargo.toml [hunk 13]
```diff
@@ -8,10 +8,18 @@ description = "Subsystem traits and message definitions"
...
+tracing = "0.1.29"
 
 metered-channel = { path = "../metered-channel" }
 
 substrate-prometheus-endpoint = { git = "https://github.com/paritytech/substrate", branch = "master" }
+sc-tracing = { git = "https://github.com/paritytech/substrate", branch = "master" }
+sc-cli = { git = "https://github.com/paritytech/substrate", branch = "master" }
+sc-service = { git = "https://github.com/paritytech/substrate", branch = "master" }
+codec = { package = "parity-scale-codec", version = "2.2.0" }
+sp-tracing = { git = "https://github.com/paritytech/substrate", branch = "master" }
```
Comment (drahnr) [inline node/metrics/Cargo.toml:20]: `sp-tracing` is in here twice

File: node/metrics/Cargo.toml (hunk 12 above)
Comment (sandreim) [reply node/metrics/Cargo.toml:22]: `sc-service`  shouldn;t be there, sc-cli is needed for `FnOnce(&mut sc_cli::LoggerBuilder, &sc_service::Configuration) -> ()`

File: node/metrics/Cargo.toml (hunk 13 above)
Comment (sandreim) [reply node/metrics/Cargo.toml:20]: actually not the same :) but only sc-tracing is needed.

File: node/metrics/Cargo.toml (hunk 12 above)
Comment (drahnr) [reply node/metrics/Cargo.toml:22]: Yikes. Well, it is what it is for now. Please add a comment to the why this is needed.

File: runtime/parachains/src/paras_inherent/mod.rs [hunk 14]
```diff
@@ -260,13 +261,18 @@ impl<T: Config> Pallet<T> {
 			parent_header,
 			mut disputes,
 		} = data;
-
-		let parent_header_hash = parent_header.hash();
+		sp_io::init_tracing();
+		let mut prefilter_weight_metric =
+			CounterVec::new("polkadot_create_inherent_prefilter_weight");
```
[the remark this answers is not in the document]
Comment (sandreim) [reply runtime/parachains/src/paras_inherent/mod.rs:266]: We don't have the information in this case, because the macro is expanded in the metrics impl file.

File: runtime/parachains/Cargo.toml (hunk 7 above)
Comment (sandreim) [reply runtime/parachains/Cargo.toml:46]: can't remove this one, it won't compile.

File: primitives/src/v1/mod.rs [hunk 15]
```diff
@@ -1452,11 +1452,13 @@ pub struct RuntimeMetricUpdate {
...
 fn vec_to_str<'a>(v: &'a Vec<u8>, default: &'static str) -> &'a str {
-	#[cfg(feature = "std")]
 	return std::str::from_utf8(v).unwrap_or(default)
+}
 
-	#[cfg(not(feature = "std"))]
+#[cfg(not(feature = "std"))]
+fn vec_to_str<'a>(v: &'a Vec<u8>, default: &'static str) -> &'a str {
 	return sp_std::str::from_utf8(v).unwrap_or(default)
 }
```
Comment (drahnr) [inline primitives/src/v1/mod.rs:1463]: Just use `sp_std`, it should internally fallback to `::std` anyways :)

File: runtime/metrics/src/lib.rs [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+		let metric_update = RuntimeMetricUpdate {
+			metric_name: sp_std::vec::Vec::from(name),
+			op: RuntimeMetricOp::Register(RuntimeMetricRegisterParams::new(
+				sp_std::vec::Vec::from(description),
+				labels,
+			)),
+		}
+		.encode();
+
+		// `from_utf8_unchecked` is safe, we only care about the `metric_name` which is static str.
```
Comment (pepyakin) [inline runtime/metrics/src/lib.rs:61]: It is not obvious for me why this is safe. More specifically, why we can be sure that in a serialized representation of `RuntimeMetricUpdate` there will be no invalid UTF-8 sequences. At least we should update the safety proof.

However, I am wondering, why don't we use a safer version for this?

File: runtime/metrics/src/lib.rs [hunk 17]
```diff
@@ -0,0 +1,128 @@
...
+
+	/// Increment by `value`.
+	pub fn inc_by(&mut self, value: u64) {
+		let metric_update = RuntimeMetricUpdate {
+			metric_name: sp_std::vec::Vec::from(self.name),
+			op: RuntimeMetricOp::Increment(value, self.label_values.clone()),
+		}
+		.encode();
+
+		// `from_utf8_unchecked` is safe, we only care about the `metric_name` which is static str.
```
Comment (pepyakin) [inline runtime/metrics/src/lib.rs:90]: same as above

File: runtime/metrics/src/lib.rs (hunk 16 above)
Comment (drahnr) [reply runtime/metrics/src/lib.rs:61]: Technically we could reduce this to be byte strings, and only try to parse on the node side into valid utf8 `String` types.

Comment (sandreim) [reply runtime/metrics/src/lib.rs:61]: We can be certain that a serialized representation of RuntimeMetricUpdate will not be a valid UTF-8 sequence. Actually, what we want si to `transmute` from u8 slice to str.

Comment (sandreim) [reply runtime/metrics/src/lib.rs:61]: I will update the comments.

File: runtime/metrics/src/nostd.rs [hunk 18]
```diff
@@ -0,0 +1,112 @@
...
+}
+
+fn emit_metric(metric_op: &[u8]) {
+	// `from_utf8_unchecked` is safe in this context:
+	// We cannot guarantee that the input slice is a valid UTF-8, but the `bs58` encoding
+	// provides that guarantee forward. We just need to `transmute` as the layout of `&str`
+	// and `&[u8]` is the same.
+
+	unsafe {
+		let metric_op = bs58::encode(sp_std::str::from_utf8_unchecked(&metric_op)).into_string();
```
Comment (ordian) [inline runtime/metrics/src/nostd.rs:26]: the only unsafe function here is `sp_std::str::from_utf8_unchecked(&metric_op)`, where `metric_op` is input `&[u8]`, so I don't get the safety proof and the reason to use this function in the first place? premature optimization?

File: runtime/parachains/src/paras_inherent/mod.rs [hunk 19]
```diff
@@ -260,22 +261,51 @@ impl<T: Config> Pallet<T> {
 			parent_header,
 			mut disputes,
 		} = data;
+		sp_io::init_tracing();
+		let mut weight_metric = CounterVec::new(
+			"parachain_inherent_data_weight",
+			"Inherent data weight before and after filtering",
+			&vec!["when"],
```
Comment (ordian) [inline runtime/parachains/src/paras_inherent/mod.rs:268]: any reason to use a `vec!` and allocate htere and elsewhere? othwerwise, `s/&vec![/&[/g`
```suggestion
			&["when"],
```

File: runtime/metrics/src/nostd.rs (hunk 18 above)
Comment (drahnr) [reply runtime/metrics/src/nostd.rs:26]: Imho we should still use a bytestring / byteslice for the key and avoid all of the string based stuff in the runtime

Comment (sandreim) [reply runtime/metrics/src/nostd.rs:26]: This is not an optimization in any case, it's just making sure our encoded metric updates comply with the tracing interfaces - the tracing API which requires that fields of Events/Spans are all Strings or integers. It doesn't allow any other types, such as the ones we are interested in.

I don't think we can avoid this, unless we make changes to the tracing API.  

In our specific context, all of the `Event` parameters are further formatted into a single string and handled in the `TraceHandler` impl (they are string values).

Does this make sense ?

Comment (ordian) [reply runtime/metrics/src/nostd.rs:26]: `bs58::encode` accepts `AsRef<[u8]>`: https://docs.rs/bs58/0.4.0/bs58/fn.encode.html
why do we cast it to `&str` then?

Comment (drahnr) [reply runtime/metrics/src/nostd.rs:26]: In that case, you are correct, and we can't.

Comment (sandreim) [reply runtime/metrics/src/nostd.rs:26]: Yeah, got it, it makes sense.

File: runtime/parachains/src/paras_inherent/mod.rs [hunk 20]
```diff
@@ -333,11 +367,23 @@ impl<T: Config> Pallet<T> {
 			// Note that `provide_multi_dispute_data` will iterate, verify, and import each
 			// dispute; so the input here must be reasonably bounded.
 			let _ = T::DisputesHandler::provide_multi_dispute_data(disputes.clone())?;
+			dispute_sets_processed_metric
+				.with_label_values(&vec!["imported)"])
```
Comment (ordian) [inline runtime/parachains/src/paras_inherent/mod.rs:371]: is closing brace `)` intended here?

File: runtime/metrics/src/nostd.rs (hunk 18 above)
Comment (sandreim) [reply runtime/metrics/src/nostd.rs:26]: I missed the fact that it can bs58 take u8 slice as input.

File: runtime/metrics/Cargo.toml [hunk 21]
```diff
@@ -0,0 +1,25 @@
+[package]
+name = "polkadot-runtime-metrics"
+version = "0.9.13"
+authors = ["Parity Technologies <admin@parity.io>"]
+edition = "2018"
```
Comment (ordian) [inline runtime/metrics/Cargo.toml:5]:
```suggestion
edition = "2021"
```

Comment (ordian) [reply runtime/metrics/Cargo.toml:5]: nvm, #4279 isn't merged yet

File: cli/src/command.rs [hunk 22]
```diff
@@ -267,12 +276,17 @@ fn run_node_inner(cli: Cli, overseer_gen: impl service::OverseerGen) -> Result<(
 	})
 }
 
+/// TODO: Start the node with support for publishing runtime metrics.
```
Comment (ordian) [inline cli/src/command.rs:279]: still unresolved?

Comment (sandreim) [reply cli/src/command.rs:279]: This needs to be removed.

File: runtime/parachains/src/paras_inherent/mod.rs (hunk 20 above)
Comment (sandreim) [reply runtime/parachains/src/paras_inherent/mod.rs:371]: No, thanks for spotting this typo/copy paster error.

File: runtime/parachains/src/paras_inherent/mod.rs [hunk 23]
```diff
@@ -333,11 +367,23 @@ impl<T: Config> Pallet<T> {
...
+				// Relay chain freeze, at this point we will not include any parachain blocks.
+				dispute_sets_processed_metric.with_label_values(&vec!["frozen)"]).inc_by(1);
+
 				// The relay chain we are currently on is invalid. Proceed no further on parachains.
 				return Ok(Some(dispute_statements_weight::<T>(&disputes)).into())
 			}
 
+			// Process the dispute sets of the current session.
+			dispute_sets_processed_metric
+				.with_label_values(&vec!["current_session)"])
```
Comment (ordian) [inline runtime/parachains/src/paras_inherent/mod.rs:384]: we only increment the counter, so current_session will track all sessions?

**Review by ordian — COMMENTED** [review]
I don't have a list of useful metrics to track in the runtime, but we can start here and grow it later.

File: runtime/metrics/Cargo.toml [hunk 24]
```diff
@@ -0,0 +1,25 @@
...
+sp-std = { package = "sp-std", git = "https://github.com/paritytech/substrate", branch = "master", default-features = false}
+sp-tracing = { git = "https://github.com/paritytech/substrate", branch = "master", default-features = false }
+parity-scale-codec = { version = "2.3.1", default-features = false }
+primitives = { package = "polkadot-primitives", path = "../../primitives", default-features = false }
+
+bs58 = { version = "0.4.0", default-features = false, features = ["alloc"] }
+
+[features]
+default = ["std"]
+no_std = []
```
Comment (ordian) [inline runtime/metrics/Cargo.toml:17]: no_std shouldn't be a feature

File: runtime/parachains/src/paras_inherent/mod.rs (hunk 23 above)
Comment (sandreim) [reply runtime/parachains/src/paras_inherent/mod.rs]: Yes, it will track all sessions. I'm unsure this is useful, I guess it depends on how long is one session ?

File: runtime/parachains/src/paras_inherent/mod.rs [hunk 25]
```diff
@@ -333,11 +367,23 @@ impl<T: Config> Pallet<T> {
 			// Note that `provide_multi_dispute_data` will iterate, verify, and import each
 			// dispute; so the input here must be reasonably bounded.
 			let _ = T::DisputesHandler::provide_multi_dispute_data(disputes.clone())?;
+			dispute_sets_processed_metric
+				.with_label_values(&["imported"])
+				.inc_by(disputes.len() as u64);
+
 			if T::DisputesHandler::is_frozen() {
+				// Relay chain freeze, at this point we will not include any parachain blocks.
+				dispute_sets_processed_metric.with_label_values(&["frozen)"]).inc_by(1);
```
Comment (ordian) [inline runtime/parachains/src/paras_inherent/mod.rs:376]: some leftovers
```suggestion
				dispute_sets_processed_metric.with_label_values(&["frozen"]).inc_by(1);
```

File: runtime/parachains/src/paras_inherent/mod.rs [hunk 26]
```diff
@@ -347,11 +393,22 @@ impl<T: Config> Pallet<T> {
 					.map(|(_, candidate)| *candidate)
 					.collect::<BTreeSet<CandidateHash>>();
 
-				let freed_disputed =
+				// Count invalid dispute sets.
+				dispute_sets_processed_metric
+					.with_label_values(&["concluded_invalid)"])
```
Comment (ordian) [inline runtime/parachains/src/paras_inherent/mod.rs:398]:
```suggestion
					.with_label_values(&["concluded_invalid"])
```

File: runtime/parachains/src/paras_inherent/mod.rs [hunk 27]
```diff
@@ -347,11 +393,22 @@ impl<T: Config> Pallet<T> {
...
 					<inclusion::Pallet<T>>::collect_disputed(&concluded_invalid_disputes)
 						.into_iter()
 						.map(|core| (core, FreedReason::Concluded))
 						.collect();
+
+				// Count concluded disputes.
+				dispute_sets_processed_metric
+					.with_label_values(&["concluded_valid)"])
+					.inc_by(freed_disputed.len() as u64);
+
```
Comment (ordian) [inline runtime/parachains/src/paras_inherent/mod.rs:411]: I don't think the name here is correct, I'd suggest to remove this for now
```suggestion
```

File: runtime/parachains/src/paras_inherent/mod.rs [hunk 28]
```diff
@@ -333,11 +367,23 @@ impl<T: Config> Pallet<T> {
...
+				// Relay chain freeze, at this point we will not include any parachain blocks.
+				dispute_sets_processed_metric.with_label_values(&["frozen)"]).inc_by(1);
+
 				// The relay chain we are currently on is invalid. Proceed no further on parachains.
 				return Ok(Some(dispute_statements_weight::<T>(&disputes)).into())
 			}
 
+			// Process the dispute sets of the current session.
+			dispute_sets_processed_metric
+				.with_label_values(&["current_session)"])
```
Comment (ordian) [inline runtime/parachains/src/paras_inherent/mod.rs:384]: maybe `newly_processed`?
```suggestion
				.with_label_values(&["newly_processed"])
```

File: runtime/parachains/src/paras_inherent/mod.rs [hunk 29]
```diff
@@ -260,22 +261,51 @@ impl<T: Config> Pallet<T> {
 			parent_header,
 			mut disputes,
 		} = data;
+		sp_io::init_tracing();
```
Comment (drahnr) [inline runtime/parachains/src/paras_inherent/mod.rs:264]: Does this need to be guarded?

File: runtime/parachains/src/paras_inherent/mod.rs [hunk 30]
```diff
@@ -260,22 +261,51 @@ impl<T: Config> Pallet<T> {
...
+		let mut dispute_sets_processed_metric = CounterVec::new(
+			"parachain_inherent_data_dispute_sets_processed",
+			"Counts the number of dispute statements sets processed in `enter_inner`.",
+			&["category"],
+		);
+
+		let mut disputes_included_metric = Counter::new(
+			"parachain_inherent_data_disputes_included",
+			"Counts the number of dispute statements sets included in a block in `enter_inner`.",
+		);
```
Comment (drahnr) [inline runtime/parachains/src/paras_inherent/mod.rs:291]: This is quite a few loc.

**Review by drahnr — COMMENTED** [review]
A few nits, almost there :)

File: primitives/src/v1/metrics.rs [hunk 31]
```diff
@@ -0,0 +1,121 @@
...
+
+	/// Returns a label names as an `Option` of `Vec<&str>`.
+	pub fn labels(&self) -> Option<Vec<&str>> {
+		self.labels.as_ref().map(|labels| labels.as_str())
+	}
+}
+
+impl RuntimeMetricLabels {
+	/// Returns a labels as `Vec<&str>`.
+	pub fn as_str(&self) -> Vec<&str> {
```
[the remark this answers is not in the document]
Comment (sandreim) [reply primitives/src/v1/metrics.rs:61]: yep, missed to update this one.

File: runtime/parachains/src/paras_inherent/mod.rs (hunk 29 above)
Comment (sandreim) [reply runtime/parachains/src/paras_inherent/mod.rs:264]: No, we need to call it once before all else, IIUC.

File: runtime/parachains/src/paras_inherent/mod.rs (hunk 28 above)
Comment (sandreim) [reply runtime/parachains/src/paras_inherent/mod.rs:384]: I would go for `current` to differentiate vs `imported`.

File: node/metrics/src/runtime.rs [hunk 32]
```diff
@@ -0,0 +1,210 @@
...
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Runtime Metrics helpers.
+//!
+//! A runtime metric provider implementation that builds on top of Substrate wasm
+//! tracing support. This requires that the custom profiler (`TraceHandler`) to be
+//! registered in substrate via a `logger_hook()`. Events emitted from runtime are
+//! then captured/processed by the `TraceHandler` implementation.
+#![cfg(feature = "runtime-metrics")]
```
Comment (drahnr) [inline node/metrics/src/runtime.rs:23]: :+1:

File: node/metrics/src/runtime.rs [hunk 33]
```diff
@@ -0,0 +1,210 @@
...
+				.and_modify(|counter_vec| counter_vec.inc_by(value));
+			Ok(())
+		})
+	}
+
+	fn with_counters_lock_held<F>(&self, do_something: F)
+	where
+		F: FnOnce(MutexGuard<'_, HashMap<String, Counter<U64>>>) -> Result<(), PrometheusError>,
+	{
+		let _ = self.1.counters.lock().map(do_something).or_else(|error| {
```
Comment (drahnr) [inline node/metrics/src/runtime.rs:106]: nit: `map_err` is a mild improvement

File: node/metrics/src/tests.rs [hunk 34]
```diff
@@ -0,0 +1,70 @@
...
+			hyper::body::to_bytes(res).await.expect("can't get body as bytes").to_vec(),
+		)
+		.expect("body is not an UTF8 string");
+
+		// Time to die.
+		kill(Pid::from_raw(cmd.id().try_into().unwrap()), SIGINT)
+			.expect("failed to kill the node process");
+
+		// If the node has authored at least 1 block this should pass.
+		assert!(body.contains(&RUNTIME_METRIC_NAME));
```
Comment (drahnr) [inline node/metrics/src/tests.rs:68]: :+1:

File: runtime/parachains/src/paras_inherent/mod.rs (hunk 29 above)
Comment (drahnr) [reply runtime/parachains/src/paras_inherent/mod.rs:264]: I mean guaraded by a feature flag?

File: runtime/parachains/src/paras_inherent/mod.rs [hunk 35]
```diff
@@ -260,13 +261,38 @@ impl<T: Config> Pallet<T> {
...
 
 		log::debug!(
 			target: LOG_TARGET,
-			"[enter_inner] parent_header={:?} bitfields.len(): {}, backed_candidates.len(): {}, disputes.len(): {}",
-			parent_header_hash,
+			"[enter] bitfields.len(): {}, backed_candidates.len(): {}, disputes.len() {}",
```
Comment (drahnr) [inline runtime/parachains/src/paras_inherent/mod.rs:295]: Why is this removed? That was explicitly added since that helps to lookup the block in the UI

File: scripts/gitlab/test_linux_stable.sh [hunk 36]
```diff
@@ -5,4 +5,4 @@ set -e
 source "$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )/../common/lib.sh"
 
 time cargo test --release --locked -p polkadot-node-core-dispute-coordinator --features disputes
-time cargo test --workspace --release --verbose --locked --features=runtime-benchmarks
+time cargo test --workspace --release --verbose --locked --features=runtime-benchmarks,runtime-metrics
```
Comment (drahnr) [inline scripts/gitlab/test_linux_stable.sh:8]: We probably want to test both variants, right?

**Review by drahnr — APPROVED** [review]
LGTM, a few small nits, but looks good overall. Encapsulating the `metrics` boilerplate is going to be addressed in a follow up PR.

File: scripts/gitlab/test_linux_stable.sh (hunk 36 above)
Comment (sandreim) [reply scripts/gitlab/test_linux_stable.sh:8]: Yeah, but that'd probably double the pipeline duration.

File: runtime/parachains/src/paras_inherent/mod.rs (hunk 35 above)
Comment (sandreim) [reply runtime/parachains/src/paras_inherent/mod.rs:295]: I don't know, seems to be an error on my side.

Comment (sandreim) [reply runtime/parachains/src/paras_inherent/mod.rs:295]: I guess this is merge damage. I'll fix it.

File: scripts/gitlab/test_linux_stable.sh (hunk 36 above)
Comment (drahnr) [reply scripts/gitlab/test_linux_stable.sh:8]: I wouldn't think so. We should definitely check that the release-able version is good at all times.

The `feature=disputes` will be removed soon, so this should be fine for the time being.

File: runtime/parachains/src/paras_inherent/mod.rs (hunk 29 above)
Comment (sandreim) [reply runtime/parachains/src/paras_inherent/mod.rs:264]: It is a no-op in the native runtime case or if `with-tracing` feature is not enabled in substrate.

File: runtime/parachains/src/paras_inherent/mod.rs [hunk 37]
```diff
@@ -333,11 +364,23 @@ impl<T: Config> Pallet<T> {
...
+				// Relay chain freeze, at this point we will not include any parachain blocks.
+				dispute_sets_processed_metric.with_label_values(&["frozen"]).inc();
+
 				// The relay chain we are currently on is invalid. Proceed no further on parachains.
 				return Ok(Some(dispute_statements_weight::<T>(&disputes)).into())
 			}
 
+			// Process the dispute sets of the current session.
+			dispute_sets_processed_metric
+				.with_label_values(&["current)"])
```
Comment (ordian) [inline runtime/parachains/src/paras_inherent/mod.rs:381]: although I'm not a fan of `current`, because we accumulate for all sessions, but it's better than `current)` :)
```suggestion
				.with_label_values(&["current"])
```

**Review by ordian — APPROVED** [review]

File: runtime/parachains/src/paras_inherent/mod.rs (hunk 37 above)
Comment (sandreim) [reply runtime/parachains/src/paras_inherent/mod.rs:381]: Damn, I never see them typos 🙄 . I think the rate statistic will still provide some value. I'll remove/update metrics in followup PR where I'll also implement a boilerplate for making use of these metrics.

File: scripts/gitlab/test_linux_stable.sh [hunk 38]
```diff
@@ -6,3 +6,5 @@ source "$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )/../com
 
 time cargo test --release --locked -p polkadot-node-core-dispute-coordinator --features disputes
 time cargo test --workspace --release --verbose --locked --features=runtime-benchmarks
+# Builds with the runtime metrics feature are only to be used for testing.
+time cargo test --workspace --release --verbose --locked --features=runtime-metrics
```
Comment (TriplEight) [inline scripts/gitlab/test_linux_stable.sh:10]: can the feature be added to the previous command?

Comment (ordian) [reply scripts/gitlab/test_linux_stable.sh:10]: see a comment above: https://github.com/paritytech/polkadot/pull/4483#discussion_r769590121

File: scripts/gitlab/test_linux_stable.sh (hunk 36 above)
Comment (drahnr) [reply scripts/gitlab/test_linux_stable.sh]: Hmm, I think @andronik has a point

**Review by TriplEight — APPROVED** [review]

File: runtime/metrics/src/std.rs [hunk 39]
```diff
@@ -0,0 +1,42 @@
...
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Runtime metrics are wasm only so we also need to provide a dummy metric
+//! implementation for the native runtime. These are required to avoid cluttering
+//! the code with `#[cfg(not(feature = "std"))]` for updating metrics.
+pub struct Counter;
```
Comment (bkchr) [inline runtime/metrics/src/std.rs:20]:
```suggestion

pub struct Counter;
```

Comment (bkchr) [reply runtime/metrics/src/std.rs:20]: Docs are missing here and for the other types, also in nostd.

File: runtime/metrics/src/lib.rs [hunk 40]
```diff
@@ -0,0 +1,33 @@
...
+
+#[cfg(feature = "std")]
+mod std;
+#[cfg(feature = "std")]
+pub use crate::std::*;
+
+#[cfg(not(feature = "std"))]
+mod nostd;
+#[cfg(not(feature = "std"))]
+pub use crate::nostd::*;
```
Comment (bkchr) [inline runtime/metrics/src/lib.rs:33]: This should be changed.

This code currently means, even if we don't have `runtime-metrics` enabled, we will use the `nostd` version that is not a `no-op` when `runtime-metrics` are disabled.

The files should be renamed to `with_runtime_metrics` and `without_runtime_metrics` to reflect this better and then the file should be chose based on the `runtime-metrics` feature.

The code should be a no-op for production.

File: runtime/parachains/src/paras_inherent/mod.rs (hunk 29 above)
Comment (drahnr) [reply runtime/parachains/src/paras_inherent/mod.rs:264]: I'd prefer to guard this here, so we are not surprised in case something in substrate changes.

File: runtime/metrics/src/std.rs (hunk 39 above)
Comment (sandreim) [reply runtime/metrics/src/std.rs]: fixed

File: runtime/metrics/src/lib.rs (hunk 40 above)
Comment (sandreim) [reply runtime/metrics/src/lib.rs]: fixed

File: primitives/src/v1/metrics.rs [hunk 41]
```diff
@@ -0,0 +1,121 @@
+use parity_scale_codec::{Decode, Encode};
```
Comment (bkchr) [inline primitives/src/v1/metrics.rs:1]: Missing copyright header.

**Review by bkchr — COMMENTED** [review]
One last nitpick

File: primitives/src/v1/metrics.rs (hunk 41 above)
Comment (sandreim) [reply primitives/src/v1/metrics.rs:1]: added.
